p2p-tests/lib/p2p.ts

```diff
@@ -104,66 +86,151 @@ export class LinePrefix extends stream.Transform {
+
+  constructor({ namespace, ipAddress, startArgs, dataPath }: PeerParams) {
+    this.dataPath = dataPath;
+    this.radHome = path.join(dataPath, `${namespace}-rad-home`);
+    this.namespace = namespace;
+    this.ipAddress = ipAddress;
+    this.startArgs = startArgs;
+
+    fs.mkdirsSync(this.radHome);
+  }
+
+  get childProcess() {
+    if (this._childProcess) {
+      return this._childProcess;
+    } else {
+      throw new Error(
+        "Tried to access childProcess before it was initialized."
+      );
     }
-  );
+  }
 
-  processes.push(childProcess);
+  async stop(): Promise<void> {
+    if (!this.childProcess) {
+      throw new Error("Tried to stop() process that wasn't started.");
+    }
+
+    this.childProcess.kill("SIGTERM");
+    await this.childProcess;
+  }
 
-  return new ProxyClient.ProxyClient(`http://${ip}:17246`);
+  async kill(): Promise<void> {
+    if (!this.childProcess) {
+      throw new Error("Tried to kill() process that wasn't started.");
+    }
+
+    try {
+      this.childProcess.kill("SIGKILL");
+      await this.childProcess;
+    } catch {
+      // Ignore errors.
+    }
+  }
 }
 
-export function startSeed(namespace: string, args: string[]): void {
-  const radHome = workspacePath([namespace]);
-  fs.removeSync(radHome);
-  fs.mkdirsSync(radHome);
+export class UpstreamPeer extends Peer {
+  _proxyClient: ProxyClient.ProxyClient | undefined = undefined;
+  public checkoutPath: string;
+  public passphrase: string;
 
-  const childProcess = spawnInNamespace(
-    namespace,
-    [path.join(BIN_PATH, "upstream-seed"), "--rad-home", radHome, ...args],
-    {}
-  );
+  constructor({ namespace, ipAddress, startArgs, dataPath }: PeerParams) {
+    super({ namespace, ipAddress, startArgs, dataPath });
+    this.checkoutPath = path.join();
+    this.passphrase = namespace;
+
+    this.init();
+  }
+
+  init(): void {
+    execa.sync(path.join(BIN_PATH, "radicle-proxy-init"), [
+      this.namespace,
+      "--key-passphrase",
+      this.passphrase,
+      "--rad-home",
+      this.radHome,
+    ]);
+  }
+
+  get proxyClient() {
+    if (this._proxyClient) {
+      return this._proxyClient;
+    } else {
+      throw new Error("Tried to access proxyClient before it was initialized.");
+    }
+  }
+
+  start(): void {
+    const childProcess = spawnInNamespace(
+      this.namespace,
+      [
+        path.join(BIN_PATH, "radicle-proxy"),
+        "--peer-listen",
+        `${this.ipAddress}:8776`,
+        "--http-listen",
+        `${this.ipAddress}:17246`,
+        "--key-passphrase",
+        this.passphrase,
+        "--skip-remote-helper-install",
+        "--unsafe-fast-keystore",
+        "--insecure-http-api",
+        "--dev-log",
+        ...this.startArgs,
+      ],
+      {
+        RAD_HOME: this.radHome,
+      }
+    );
+
+    this._childProcess = childProcess;
+    this._proxyClient = new ProxyClient.ProxyClient(
+      `http://${this.ipAddress}:17246`
+    );
```

Comment:
I’d suggest moving this to the constructor. This way, `this._proxyClient` cannot be `undefined` and we can eliminate the `proxyClient()` getter.

Reply:
c3f5d7cffee1dc92cf962eab929bd11db8cddd67